R/driver-sql-server.R

```diff
@@ -81,6 +81,46 @@ setMethod("dbExistsTable", c("Microsoft SQL Server", "character"),
   }
 )
 
+#' @description
+#' ## `dbListTables()`
+#' The default implementation reports temporary tables as non-existent
+#' when a `catalog_name` isn't supplied since they live in a different catalog.
+#' This method provides a special case for temporary tables.
+#' @rdname SQLServer
+#' @usage NULL
+setMethod("dbListTables", "Microsoft SQL Server",
+  function(conn,
+           catalog_name = NULL,
+           schema_name = NULL,
+           table_name = NULL,
+           table_type = NULL,
+           ...) {
+    res <- callNextMethod(
+      conn = conn,
+      catalog_name = catalog_name,
+      schema_name = schema_name,
+      table_name = table_name,
+      table_type = table_type,
+      ...
+    )
+
+    if (is.null(catalog_name)) {
+      res_temp <- callNextMethod(
+        conn = conn,
+        catalog_name = "tempdb",
+        schema_name = schema_name,
+        table_name = table_name,
+        table_type = table_type,
+        ...
+      )
```

Comment:
```suggestion
      res_temp <- callNextMethod(
        catalog_name = "tempdb",
      )
```

I am no `callNextMethod()` expert but I believe the main way it works is MAGIC and you only need to supply arguments that are different to the caller.

And I think you might also need `schema_name = "dbo"` (Reading between the lines from https://stackoverflow.com/questions/8311959/specifying-schema-for-temporary-tables and my recollection of how `dbExistsTable()` was previously implemented)

Reply:
Woah😮

Works well for the first call, `callNextMethod()`, but writing `callNextMethod(catalog_name = "tempdb")` for the second results in:

```
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'conn' in selecting a method for function 'odbcConnectionTables': argument "conn" is missing, with no default
```

Seems like it needs the arguments preceding any (in position) that are manually set; adding `conn = con` and leaving the rest unspecified seems to work but will confirm.

------

re: dbo, I did notice that `tempdb` has analogous `INFORMATION_SCHEMA` and `sys` schemas to those in `master` (or any non-default catalog). We currently include tables in those schemas from those other databases by default; should we exclude them for `tempdb` by setting `schema_name = "dbo"`?

Reply:
Sorry, somehow missed this and every other comment 😬 

I am super surprised that you get that error because the docs say "the arguments have no effect on selecting the next method.", but they clearly do! But maybe `callNextMethod()` is confused because we've already called it once? I wish I understood what was happening here better, but it's probably not worth the time to explore, and if your code works, then go with it.

I do think we should set `schema_name = "dbo"` too.